src/widget/wvumetergl.h

```diff
@@ -72,4 +80,15 @@ class WVuMeterGL : public QGLWidget, public WBaseWidget {
     double m_dPeakHoldCountdownMs;
 
     QColor m_qBgColor;
+#ifdef MIXXX_USE_QOPENGL
+    bool m_hasOpenGLShaderPrograms{};
+    std::unique_ptr<QOpenGLTexture> m_pTextureBack;
+    std::unique_ptr<QOpenGLTexture> m_pTextureVu;
+    QOpenGLShaderProgram m_shaderProgram;
+
+    void initializeGL() override;
+    void cleanupGL();
+    void renderGL() override;
+    void drawTexture(QOpenGLTexture* texture, const QRectF& sourceRect, const QRectF& targetRect);
+#endif
```

Comment:
So previously the `WVuMeterGL` didn't use GL? Or OpenGL? I don't quite understand this change here (same for `WSpinny`). Is there a way to do this less intrusively?

Reply:
Correct,  previously, or now without MIXXX_USE_QOPENGL, they would use a QPainter on the OpenGL paint device, which worked kind of okay with QGLWidget, but results to be very inefficient with the QOpenGLWindow based solution.

I guess a cleaner solution could be to have a base class with the common code, and derive both the qpainter version and the pure qopengl version from that base class. Same for spinny.

Reply:
Ok, after looking at the code I remember why I did it like this. In the case of MIXXX_USE_QOPENGL I also keep the QPainter code, to use as a fallback in case GLSL is not supported on the system. So this actually does suggest that a situation exists where both the QPainter and the GLSL version coexist (and the GLSL version is used if the system supports GLSL, otherwise fall back to the QPainter version). I will make the change accordingly.

Reply:
Update: The first debug assertion is triggered ~10 times, finally followed by the `reserved == ...` assertion after repeatedly continuing execution:

```
critical [Main] DEBUG ASSERT: "it == cbegin() || it == cend() || (*it >= position && *std::prev(it) < position)" in function mixxx::Beats::ConstIterator mixxx::Beats::iteratorFrom(mixxx::audio::FramePos) const at /home/uk/volumes/Build/cpp/mixxx/src/track/beats.cpp:454
Process 76706 stopped
* thread #1, name = 'mixxx', stop reason = signal SIGINT
    frame #0: 0x00007ffff2aafb94 libc.so.6`__pthread_kill_implementation + 276
libc.so.6`__pthread_kill_implementation:
->  0x7ffff2aafb94 <+276>: movl   %eax, %ebx
    0x7ffff2aafb96 <+278>: negl   %ebx
    0x7ffff2aafb98 <+280>: cmpl   $0xfffff000, %eax         ; imm = 0xFFFFF000 
    0x7ffff2aafb9d <+285>: movl   $0x0, %eax
(lldb) c
Process 76706 resuming
critical [Main] DEBUG ASSERT: "reserved == m_vertices.size()" in function virtual void qopengl::WaveformRenderBeat::renderGL() at /home/uk/volumes/Build/cpp/mixxx/src/waveform/renderers/qopengl/waveformrenderbeat.cpp:102
Process 76706 stopped
* thread #1, name = 'mixxx', stop reason = signal SIGINT
    frame #0: 0x00007ffff2aafb94 libc.so.6`__pthread_kill_implementation + 276
libc.so.6`__pthread_kill_implementation:
->  0x7ffff2aafb94 <+276>: movl   %eax, %ebx
    0x7ffff2aafb96 <+278>: negl   %ebx
    0x7ffff2aafb98 <+280>: cmpl   $0xfffff000, %eax         ; imm = 0xFFFFF000 
    0x7ffff2aafb9d <+285>: movl   $0x0, %eax
```

Forgot to create a backtrace when it happened. Will do so next time.

Reply:
```
critical [Main] DEBUG ASSERT: "it == cbegin() || it == cend() || (*it >= position && *std::prev(it) < position)" in function mixxx::Beats::ConstIterator mixxx::Beats::iteratorFrom(mixxx::audio::FramePos) const at /home/uk/volumes/Build/cpp/mixxx/src/track/beats.cpp:454
Process 83843 stopped
* thread #1, name = 'mixxx', stop reason = signal SIGINT
    frame #0: 0x00007ffff2aafb94 libc.so.6`__pthread_kill_implementation + 276
libc.so.6`__pthread_kill_implementation:
->  0x7ffff2aafb94 <+276>: movl   %eax, %ebx
    0x7ffff2aafb96 <+278>: negl   %ebx
    0x7ffff2aafb98 <+280>: cmpl   $0xfffff000, %eax         ; imm = 0xFFFFF000 
    0x7ffff2aafb9d <+285>: movl   $0x0, %eax
(lldb) bt
* thread #1, name = 'mixxx', stop reason = signal SIGINT
  * frame #0: 0x00007ffff2aafb94 libc.so.6`__pthread_kill_implementation + 276
    frame #1: 0x00007ffff2a5eaee libc.so.6`raise + 30
    frame #2: 0x0000000000683a01 mixxx`mixxx::(anonymous namespace)::handleMessage(QtMsgType, QMessageLogContext const&, QString const&) at logging.cpp:346:18
    frame #3: 0x00007ffff30ecb88 libQt5Core.so.5`qt_message_print(QtMsgType, QMessageLogContext const&, QString const&) + 280
    frame #4: 0x00007ffff30eccae libQt5Core.so.5`qt_message(QtMsgType, QMessageLogContext const&, char const*, __va_list_tag*) + 62
    frame #5: 0x00007ffff30b144b libQt5Core.so.5`QMessageLogger::critical(char const*, ...) const + 181
    frame #6: 0x00000000004a5c76 mixxx`mixxx_debug_assert(assertion=<unavailable>, file=<unavailable>, line=<unavailable>, function=<unavailable>) at assert.h:9:14
    frame #7: 0x0000000000513a97 mixxx`mixxx::Beats::iteratorFrom(mixxx::audio::FramePos) const (.cold) at beats.cpp:454:5
    frame #8: 0x00000000009cabde mixxx`mixxx::Beats::numBeatsInRange(mixxx::audio::FramePos, mixxx::audio::FramePos) const at beats.cpp:464:27
    frame #9: 0x0000000000ac3676 mixxx`qopengl::WaveformRenderBeat::renderGL() at waveformrenderbeat.cpp:71:60
    frame #10: 0x00000000006c8770 mixxx`qopengl::WaveformWidget::renderGL() at waveformwidget.cpp:36:68
    frame #11: 0x00000000006c86d2 mixxx`non-virtual thunk to qopengl::WaveformWidget::render() at waveformwidget.h:20:21
    frame #12: 0x00000000006adf40 mixxx`WaveformWidgetFactory::render() at waveformwidgetfactory.cpp:725:40
    frame #13: 0x00007ffff32df5db libQt5Core.so.5`QObject::event(QEvent*) + 539
```

Reply:
After ~10 repetitions finally followed by:

```
critical [Main] DEBUG ASSERT: "reserved == m_vertices.size()" in function virtual void qopengl::WaveformRenderBeat::renderGL() at /home/uk/volumes/Build/cpp/mixxx/src/waveform/renderers/qopengl/waveformrenderbeat.cpp:102
Process 83843 stopped
* thread #1, name = 'mixxx', stop reason = signal SIGINT
    frame #0: 0x00007ffff2aafb94 libc.so.6`__pthread_kill_implementation + 276
libc.so.6`__pthread_kill_implementation:
->  0x7ffff2aafb94 <+276>: movl   %eax, %ebx
    0x7ffff2aafb96 <+278>: negl   %ebx
    0x7ffff2aafb98 <+280>: cmpl   $0xfffff000, %eax         ; imm = 0xFFFFF000 
    0x7ffff2aafb9d <+285>: movl   $0x0, %eax
(lldb) bt
* thread #1, name = 'mixxx', stop reason = signal SIGINT
  * frame #0: 0x00007ffff2aafb94 libc.so.6`__pthread_kill_implementation + 276
    frame #1: 0x00007ffff2a5eaee libc.so.6`raise + 30
    frame #2: 0x0000000000683a01 mixxx`mixxx::(anonymous namespace)::handleMessage(QtMsgType, QMessageLogContext const&, QString const&) at logging.cpp:346:18
    frame #3: 0x00007ffff30ecb88 libQt5Core.so.5`qt_message_print(QtMsgType, QMessageLogContext const&, QString const&) + 280
    frame #4: 0x00007ffff30eccae libQt5Core.so.5`qt_message(QtMsgType, QMessageLogContext const&, char const*, __va_list_tag*) + 62
    frame #5: 0x00007ffff30b144b libQt5Core.so.5`QMessageLogger::critical(char const*, ...) const + 181
    frame #6: 0x00000000004a5c76 mixxx`mixxx_debug_assert(assertion=<unavailable>, file=<unavailable>, line=<unavailable>, function=<unavailable>) at assert.h:9:14
    frame #7: 0x0000000000531a38 mixxx`qopengl::WaveformRenderBeat::renderGL() (.cold) at waveformrenderbeat.cpp:102:5
    frame #8: 0x00000000006c8770 mixxx`qopengl::WaveformWidget::renderGL() at waveformwidget.cpp:36:68
    frame #9: 0x00000000006c86d2 mixxx`non-virtual thunk to qopengl::WaveformWidget::render() at waveformwidget.h:20:21
    frame #10: 0x00000000006adf40 mixxx`WaveformWidgetFactory::render() at waveformwidgetfactory.cpp:725:40
    frame #11: 0x00007ffff32df5db libQt5Core.so.5`QObject::event(QEvent*) + 539
```

Reply:
Thanks for the detailed report, @uklotzde , I will look into this.

Reply:
@uklotzde is this with a specific track? I can't reproduce this.

Reply:
I have submitted a possible fix. It's a bit of a guess, as I can't reproduce the issue, but I see that elsewhere the arguments for Beats::numBeatsInRange(startPos, endPos) are validated before the call. so I am doing the same now.

(I think it would be safer to do this validation inside numBeatsInRange, but first let's see if this solves the issue)

Reply:
So my possible fix didn't make a difference?

Reply:
@Swiftb0y I have restructured the WSpinny and WVuMeter classes a bit:

WSpinnyBase with derived WSpinny (which draws using QPainter) and WSpinnyGLSL (which draws using shaders). WSpinnyGLSL is used by default, unless there is no GLSL support or when QOPENGL=OFF.

WVumeterBase with derived WVumeter (which draws uses QPainter) and WVumeterGLSL (which draws using shaders). WVumeterGLSL is used by default, unless there is no GLSL support or when QOPENGL=OFF.

(Note that I have renamed WVuMeterGL to WVuMeter (so the naming scheme matches WSpinny), and the old WVuMeter (which doesn't use the WGLWidget as basis, but standard QWidget updates and thus hogs the event loop on macOS) to WVuMeterLegacy)